### PR TITLE
Double hop authentication

### PR DESCRIPTION
**Pull request by koikonom**
This PR builds on #107 to make double hop authentication work for all cases.

Also took the opportunity to refactor the way we generate powershell commands so we can account for more possible variations.

### CODE REVIEW

File: ad/internal/config/config.go [hunk 1]
```diff
@@ -247,12 +255,104 @@ func (c *KerberosTransporter) Post(_ *winrm.Client, request *soap.SoapMessage) (
 	defer resp.Body.Close()
 
 	if resp.StatusCode != 200 {
-		return "", fmt.Errorf("http error while making kerberos authenticated winRM request: %s", err)
+		return "", fmt.Errorf("http error while making kerberos authenticated winRM request: %d - %s ", resp.StatusCode, resp.Status)
```
Comment (appilon) [inline ad/internal/config/config.go:258]: including response body in error could be helpful additional context?

File: ad/internal/config/config.go [hunk 2]
```diff
@@ -247,12 +255,104 @@ func (c *KerberosTransporter) Post(_ *winrm.Client, request *soap.SoapMessage) (
...
 	body, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
 		return "", err
 	}
+
 	return string(body), err
 }
+
+// ProviderConf holds structures that are useful to the provider at runtime.
+type ProviderConf struct {
```
Comment (appilon) [inline ad/internal/config/config.go:270]: Should there be a max amount of connections for the pool? Or not really a concern?

File: ad/internal/winrmhelper/powershell_command.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	res := PSCommand{
+		CreatePSCommandOpts: opts,
+		cmd:                 cmd,
+	}
+
+	return &res
+}
+
+// Run will run a powershell command and return the stdout and stderr
+// The output is converted to JSON if the json patameter is set to true.
```
Comment (appilon) [inline ad/internal/winrmhelper/powershell_command.go:86]:
```suggestion
// The output is converted to JSON if the json parameter is set to true.
```

File: ad/internal/winrmhelper/winrm_group_membership.go [hunk 4]
```diff
@@ -53,7 +54,9 @@ func unmarshalGroupMembership(input []byte) ([]*GroupMember, error) {
 	if err != nil {
 		return nil, err
 	}
-
+	if len(gm) > 1 && gm[0].GUID == "" {
```
Comment (appilon) [inline ad/internal/winrmhelper/winrm_group_membership.go:57]: len(gm) > 0?

**Review by appilon — CHANGES_REQUESTED** [review]
small feedback, but probably approved

File: ad/internal/config/config.go (hunk 2 above)
Comment (koikonom) [reply ad/internal/config/config.go:270]: Not a concern so far.

File: ad/internal/winrmhelper/winrm_group_membership.go (hunk 4 above)
Comment (koikonom) [reply ad/internal/winrmhelper/winrm_group_membership.go:57]: yes, good catch!

File: ad/internal/config/config.go (hunk 1 above)
Comment (koikonom) [reply ad/internal/config/config.go]: Good call.

**Review by appilon — APPROVED** [review]
LGTM!

**Pull request merged**